### PR TITLE
fs: move memset to upper lever for statfs and add f_fsid field

### PR DESCRIPTION
**Pull request by zhhyu7**
## Summary
if struct statfs add new members, such as f_fsid, no additional code changes are required. and add f_fsid field for third-party code compile, because NuttX doesn't have the device number, so we're not assigning a valid value here. just memset to zero.

## Testing
sim:local

### CODE REVIEW

File: include/sys/statfs.h [hunk 1]
```diff
@@ -114,6 +114,10 @@ struct statfs
   fsblkcnt_t f_bavail;   /* Free blocks avail to non-superuser */
   fsfilcnt_t f_files;    /* Total file nodes in the file system */
   fsfilcnt_t f_ffree;    /* Free file nodes in the file system */
+  struct
+  {
+    int val[2];
+  } f_fsid;              /* Encode device type, not yet in use */
```
Comment (acassis) [inline include/sys/statfs.h:120]: I think it is not common to have a struct defined inside other struct on NuttX, shouldn't it better to define f_fsid as f_fsid_s outside of statfs struct?

**Comment by acassis** [comment]
@patacongo I think you were planing to reply @zhhyu7, but end up editing his comment, it was difficult to understand who said what. :-)

Yes, I found some examples in wqueue.h, udp.h, etc.

But I think in this case if fsid_s is planned to be used for all filesystems, eventually we will have a function that list fs based on fs IDs, etc. So in this case fsid_s could be passed as parameter to this function.

Anyway we need to define the right place to put it

File: include/sys/statfs.h (hunk 1 above)
Comment (patacongo) [reply include/sys/statfs.h]: >  I think it is not common to have a struct defined inside other struct on NuttX, shouldn't it better to define f_fsid as f_fsid_s outside of statfs struct?

Structures defined within other structures is actually very common. There are hundreds of examples. I would consider this a standard programming practice.

>  @acassis I defined the structure in types.h without typedef. is that okay?

No. Please do not modify sys/types.h. The content of that file is controlled by POSIX standards and should contain as few non-standard types as possible!!!

**Comment by patacongo** [comment]
The f_fsid field is not standard in `struct statfs` (but is required by POSIX in` struct statvfs`).  It is NOT specifid by POSIX standards for statfs.  This is handled differently by different OSs:

https://linux.die.net/man/2/statfs:

> Nobody knows what f_fsid is supposed to contain (but see below). 
> ...
> **The f_fsid field**
>
> Solaris, Irix and POSIX have a system call  that returns a struct statvfs (defined in sys/statvfs.h containing an unsigned long f_fsid.  Linux, SunOS, HP-UX, 4.4BSD have a system call statfs() that returns a struct statfs (defined in sys/vfs.h containing a fsid_t f_fsid, where fsid_t is defined as struct { int val[2]; }. The same holds for FreeBSD, except that it uses the include file sys/mount.h.
> 
> The general idea is that f_fsid contains some random stuff such that the pair (f_fsid, ino) uniquely determines a file. Some operating systems use (a variation on) the device number, or the device number combined with the file-system type. Several OSes restrict giving out the f_fsid field to the superuser only (and zero it for unprivileged users), because this field is used in the filehandle of the file system when NFS-exported, and giving it out is a security concern.
> 
> Under some operating systems the fsid can be used as second argument to the sysfs system call.

File: include/sys/statfs.h (hunk 1 above)
Comment (zhhyu7) [reply include/sys/statfs.h]: > No. Please do not modify sys/types.h. The content of that file is controlled by POSIX standards and should contain as few non-standard types as possible!!!

@patacongo wheather I need to remove this structure from the sys/types.h file? if so, which would be the more appropriate file, I don't know much about it, thanks very much.

**Pull request merged**